### PR TITLE
WIP chore: swap PhantomJs with Puppeteer

### PR DESCRIPTION
**Pull request by jkodu**
> Note: A proposal for this piece of work is here - https://github.com/dequelabs/html-team-proposals/pull/6. This PR handles [Step 1](https://github.com/dequelabs/html-team-proposals/pull/6/files#diff-da0741cfbe1cedb0339d04635fd2fc01R36) in the proposal 

This PR does the below:
- Introduces `puppeteer` for headless testing.
- Loses `grunt-mocha`, which under the hood used `PhantomJs`.
- A custom mocha reporter is used when tests are run in headless mode, which helps collate the results from the runner.
- Clean up runner templates used for generating test html files.
- Wire up `npm script` in place of `grunt-mocha` task & update grunt tasks accordingly.
- Updates `test-webdriver` scripts to listen to right mocha results and stats.
- Fixes 2 failing tests (added notes for the same - https://github.com/dequelabs/axe-core/compare/no-phantomjs?expand=1#diff-654d769210c4882644c03b54895fb9ccR347).
- removes `phantomjs` install script used in `circleci`.
- remove `grunt-mocha` from `retireignore` and `package.json`.

Closes issue:
- - https://github.com/dequelabs/axe-core/issues/305

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**
- [ ] Follows the commit message policy, appropriate for next version
- [ ] Has documentation updated, a DU ticket, or requires no documentation change
- [ ] Includes new tests, or was unnecessary
- [ ] Code is reviewed for security by: << Name here >>

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -350,16 +346,18 @@ module.exports = function(grunt) {
 			}
 		},
 		run: {
-			npm_run_imports: {
-				cmd: 'node',
-				args: ['./build/imports-generator']
+			npmRunImportsGenerator: {
+				exec: 'npm run imports-gen'
+			},
+			npmRunTestHeadless: {
```
Comment (stephenmathieson) [inline Gruntfile.js]: Please remove this. We do not need Grunt to run npm scripts.

File: build/tasks/test-webdriver.js [hunk 2]
```diff
@@ -5,7 +5,7 @@ var WebDriver = require('selenium-webdriver');
 
 module.exports = function(grunt) {
 	/**
-	 * Keep injecting scripts until window.mochaResults is set
+	 * Keep injecting scripts until `window.__mochaResult__` is set
```
Comment (stephenmathieson) [inline build/tasks/test-webdriver.js]: Why are you changing this?

File: build/test/get-test-paths.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+const globby = require('globby');
+
+/**
+ * Get list of paths/ files to test
+ */
+const getTestPaths = async () => {
+	const host = `localhost`;
```
Comment (stephenmathieson) [inline build/test/get-test-paths.js]: Why is this a template string? Also, why is these variables necessary?

File: build/test/headless.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+		console.error(`No test paths are specified to run in headless mode.`);
+	}
+
+	/**
+	 * Launch puppeteer
+	 */
+	const browser = await puppeteer.launch({
+		headless: true,
+		args: ['--no-sandbox', '-–disable-setuid-sandbox'],
+		ignoreHTTPSErrors: true,
```
Comment (stephenmathieson) [inline build/test/headless.js]: Why are we ignoring HTTPS errors?

File: build/test/headless.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+		devtools: true
```
Comment (stephenmathieson) [inline build/test/headless.js]: Why do we want the devtools?

File: build/test/headless.js [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+			console.log(
+				`Stats: ${path} (Total: ${stats.tests}, Duration: ${
+					stats.duration
+				}, Passed: ${stats.passes}, Failed: ${stats.failures}, Pending: ${
+					stats.pending
+				}\n)`
+			);
+		} catch (error) {
+			console.log(error);
```
Comment (stephenmathieson) [inline build/test/headless.js]: Should an error running a test be considered a test failure?

File: test/integration/adapter.js [hunk 7]
```diff
@@ -1,29 +1,68 @@
...
+		}
+		var result = {};
+		Object.getOwnPropertyNames(err).forEach(function(key) {
+			result[key] = err[key];
 		});
+		return result;
+	}
+
+	function setResult(stats) {
+		!window.__mochaResult__ && (window.__mochaResult__ = stats);
```
Comment (stephenmathieson) [inline test/integration/adapter.js]: Please use an `if`. This is very cryptic.

File: test/integration/adapter.js [hunk 8]
```diff
@@ -1,29 +1,68 @@
-/*global mocha */
-var failedTests = [];
+/*global mocha, Mocha */
```
Comment (stephenmathieson) [inline test/integration/adapter.js]: What is the purpose of this file?

File: test/mocha-setup.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+ * Custom mocha reporter, invoked when running in headless mode (eg: puppeteer)
+ * @param {Object} runner mocha runner instance
+ */
+function axeHeadlessModeReporter(runner) {
+	var indent = 0;
+	var failedTests = [];
+
+	Mocha.reporters.Base.call(this, runner);
+
+	runner
```
Comment (stephenmathieson) [inline test/mocha-setup.js]: This looks duplicated from the previous file. Is that intentional?

File: test/mocha-setup.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+		}
+		var result = {};
+		Object.getOwnPropertyNames(err).forEach(function(key) {
+			result[key] = err[key];
+		});
+		return result;
+	}
+
+	function setResult(stats) {
+		!window.__mochaResult__ && (window.__mochaResult__ = stats);
```
Comment (stephenmathieson) [inline test/mocha-setup.js]: This cryptic code is copypasta. Please fix it in both places (or better yet, fix it once and `require()` it 😉).

File: test/runner.tmpl [hunk 11]
```diff
@@ -4,71 +4,67 @@
...
-
-	<body>
-		<div id="fixture"></div>
-		<div id="mocha"></div>
 		<script src="/test/testutils.js"></script>
 		<% tests.forEach(function (file) { %>
 			<script src="<%=file%>"></script>
 		<%  }); %>
+		
+		<script type="text/javascript">
```
Comment (stephenmathieson) [inline test/runner.tmpl]: Why is this here? I think it belongs in a `.js` file.

File: build/test/headless.js (hunk 4 above)
Comment (AutoSponge) [reply build/test/headless.js]: This just skips the warnings generated by using a fixture server that isn't served over TLS. It shouldn't be an issue.

File: build/test/headless.js (hunk 5 above)
Comment (AutoSponge) [reply build/test/headless.js]: devtools can't be opened in headless mode. So, one of these is getting ignored.

File: build/test/headless.js (hunk 4 above)
Comment (stephenmathieson) [reply build/test/headless.js]: Since it stands out, I think we should add a comment explaining _why_ we're doing it.

File: build/test/headless.js (hunk 5 above)
Comment (jkodu) [reply build/test/headless.js]: I used it for debugging locally & left it by mistake.

File: Gruntfile.js (hunk 1 above)
Comment (jkodu) [reply Gruntfile.js]: We do until the day we lose `Grunt` and move solely to `npm scripts`, in order to maintain the sequence in which tasks are executed. For example the test execution in headless mode depends on `build`, `testconfig`, `fixture`, `connect`, which are steps for building `axe`, generating test fixtures based on runner templates and booting a server respectively.

File: build/test/get-test-paths.js (hunk 3 above)
Comment (jkodu) [reply build/test/get-test-paths.js]: Hopefully this provides some insight - https://github.com/dequelabs/html-team-proposals/pull/6/files#diff-da0741cfbe1cedb0339d04635fd2fc01R73

File: Gruntfile.js (hunk 1 above)
Comment (WilcoFiers) [reply Gruntfile.js]: @jkodu you can put those tasks in an NPM script. I agree with Stephen. We should transition out of Grunt.

File: build/test/get-test-paths.js (hunk 3 above)
Comment (stephenmathieson) [reply build/test/get-test-paths.js]: Nope. Still don't understand. I think this could be removed.

**Comment by jkodu** [comment]
Closing this. After discussion with @WilcoFiers going to explore `karma` as against authoring our own test runner.
